### PR TITLE
Edit and delete question

### PR DESCRIPTION
**Pull request by z-rayc**
## Changes
- Open and close the modal correctly for create, edit or delete question
- Deleting a question updates the arrangement for the other questions
- Replace articles datalist with dropdown; allowing for correct default value
- Live edit image for existing question
- Add placeholder text for inputs
- Refactor to use HTMX instead of JS to remove deleted article from list
- Fixed some places where I believe `err == sql.errNoRows` was being used incorrectly. Please double check those who know better.

## Yet to do
- Visual feedback on error response from request
- Edit question's time limit seconds
- Question alternatives arrangement
  - Shuffle alternatives arrangement button
- Adding article to quiz via "edit/create question" modal should update articles list (visually)

### CODE REVIEW

**Review by Molnes — COMMENTED** [review]
Looks good to me, but @nokacper24 should probably look trough this as well, especially the SQL

File: internal/models/questions/questions.go [hunk 1]
```diff
@@ -460,7 +480,61 @@ func AddNewQuestion(db *sql.DB, question Question) error {
...
+		return err
+	}
+
+	return nil
+}
+
+// Update a question in the database.
+func UpdateQuestion(db *sql.DB, question *Question) error {
+	// Start a transaction
+	ctx := context.Background()
```
Comment (nokacper24) [inline internal/models/questions/questions.go:505]: take ctx (type `context.Context`) as parameter to this function instead.
in your handler you can get the context of a request from `c.Request().Context()`

File: internal/models/questions/questions.go [hunk 2]
```diff
@@ -478,3 +552,100 @@ func AddNewQuestion(db *sql.DB, question Question) error {
...
+	}
+
+	// Update the arrangement numbers of the remaining questions
+	// (decrement all questions' arrangement with higher arrangement number)
+	_, err = tx.Exec(
+		`UPDATE questions
+		SET arrangement = arrangement - 1
+		WHERE quiz_id = $1 AND arrangement > $2;`,
+		quizID, arrangement,
+	)
```
Comment (nokacper24) [inline internal/models/questions/questions.go:610]: nice

File: internal/models/quizzes/quizzes.go [hunk 3]
```diff
@@ -195,7 +195,7 @@ func scanQuizFromFullRow(row *sql.Row) (*Quiz, error) {
 		&quiz.Published,
 		&quiz.IsDeleted,
 	)
-	if err != nil {
+	if err == sql.ErrNoRows {
```
Comment (nokacper24) [inline internal/models/quizzes/quizzes.go:198]: again i think

File: internal/models/quizzes/quizzes.go [hunk 4]
```diff
@@ -206,7 +206,7 @@ func scanQuizFromFullRow(row *sql.Row) (*Quiz, error) {
 	}
 	quiz.ImageURL = *tempURL
 
-	if err == sql.ErrNoRows {
+	if err != nil {
```
Comment (nokacper24) [inline internal/models/quizzes/quizzes.go:209]: from what i see here  the err may come from `data_handling.ConvertNullStringToURL...`
i do believe and agree that the `sql.ErrNoRows` was out of place here, but isn't the entire second `if err` pointless? this error is already checked for on line 204. There is no new error being returned at any point after that check

File: internal/web_server/web/handlers/api/admin_api.go [hunk 5]
```diff
@@ -310,25 +318,38 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
-	if article == nil {
-		tempArticle, err := articles.GetSmpArticleByURL(articleURLString)
-		if err != nil {
-			return echo.NewHTTPError(http.StatusBadRequest, "Failed to fetch article data")
+	// Only add article to DB if it is not empty.
+	// I.e. allow for no article, but not invalid article.
+	if articleURLString != "" {
+		// Check if the article URL is in the database
+		article, err = articles.GetArticleByURL(aah.sharedData.DB, articleURL)
+		if err != nil && err != sql.ErrNoRows {
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:328]: if thereis an error OTHER than `sql.ErrNoRows`, you return status bad request? Is everything that can go wrong other than no rows the clients fault? i believe there might be a 500 (internal server error) if something goes sideways on the server side as well.
To be frank i dont know exactly what  this function does so i might be wrong here. But i think you can have better errors anyway. If matching for possible sql errors or other faults is tedious here, consider defining own error types in model and checking for those

Comment (nokacper24) [reply internal/web_server/web/handlers/api/admin_api.go:328]: on second look, dont you mean that error no rows found should do bad request (or not found), and any other should turn into 500 (ie just return in)? if so you'll need nested if, first to ckeck if err != nil, then in the body check if the error is sql.NoRows -> 400/404/whatever, and else return err

File: internal/web_server/web/handlers/api/admin_api.go [hunk 6]
```diff
@@ -310,25 +318,38 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
+			return echo.NewHTTPError(http.StatusBadRequest, "Failed to get article ID")
 		}
 
-		articles.AddArticle(aah.sharedData.DB, &tempArticle)
-		article = &tempArticle
+		// If not in DB, fetch the relevant article data and add it to the DB
+		if article == nil {
+			tempArticle, err := articles.GetSmpArticleByURL(articleURLString)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusBadRequest, "Failed to fetch article data")
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:336]: is tat a badrequest? is client at fault?

File: internal/web_server/web/handlers/api/admin_api.go [hunk 7]
```diff
@@ -345,12 +366,79 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
 
-	// Save the question to the database
-	err = questions.AddNewQuestion(aah.sharedData.DB, question)
-	if err != nil {
-		return echo.NewHTTPError(http.StatusBadRequest, "Failed to create new question")
+	// Get the question by ID from the database.
+	tempQuestion, err := questions.GetQuestionByID(aah.sharedData.DB, questionID)
+
+	// If doesn't exist in the database.
+	if err != nil && err == sql.ErrNoRows {
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:373]: what if err is not `sql.ErrNoRows`? shouldnt that result in 500?

File: internal/web_server/web/handlers/api/admin_api.go [hunk 8]
```diff
@@ -345,12 +366,79 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
+		// Save the question to the database.
+		err = questions.AddNewQuestion(aah.sharedData.DB, question)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, "Failed to create new question")
+		}
+	} else if tempQuestion.ID == questionID {
+		// If the question ID is found, update the question.
+		question.ID = questionID
+		err = questions.UpdateQuestion(aah.sharedData.DB, &question)
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:382]: check for this err?

File: internal/web_server/web/handlers/api/admin_api.go [hunk 9]
```diff
@@ -345,12 +366,79 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
+	// Get the question ID
+	questionID, err := uuid.Parse(c.QueryParam(queryParamQuestionID))
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, errorInvalidQuestionID)
+	}
+
+	// Delete the question from the database
+	err = questions.DeleteQuestionByID(aah.sharedData.DB, questionID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "Failed to delete question")
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:400]: again, i dont think that is necessarily a bad request  
I'm pretty sure sql delete does not error if there was nothing to be deleted. You may be able to see the number of rows affected in the result thing returned by sql.Exec, and you return custom error if nothing was deleted? then do a not found if there was no such id to delete, and return the general err if anything else went wrong (so it results in 500)

File: internal/web_server/web/handlers/api/admin_api.go [hunk 10]
```diff
@@ -345,12 +366,79 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
+	}
+
+	return c.NoContent(http.StatusOK)
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:403]: do http.StatusNoContent (204)  if returning no content

File: internal/web_server/web/handlers/api/admin_api.go [hunk 11]
```diff
@@ -345,12 +366,79 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
+	// Get the question ID
+	questionID, err := uuid.Parse(c.QueryParam(queryParamQuestionID))
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, errorInvalidQuestionID)
+	}
+
+	// Remove the image URL from the question
+	err = questions.RemoveImageByQuestionID(aah.sharedData.DB, &questionID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "Failed to remove question image")
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:439]: again

File: internal/web_server/web/handlers/api/admin_api.go [hunk 12]
```diff
@@ -345,12 +366,79 @@ func (aah *AdminApiHandler) createQuestion(c echo.Context) error {
...
+		return echo.NewHTTPError(http.StatusBadRequest, errorInvalidQuestionID)
+	}
+
+	// Get the new image URL
+	image := c.FormValue(dashboard_components.QuestionImageURL)
+	imageURL, _ := url.Parse(image)
+
+	// Set the image URL for the question
+	err = questions.SetImageByQuestionID(aah.sharedData.DB, &questionID, imageURL)
+	if err != nil {
```
Comment (nokacper24) [inline internal/web_server/web/handlers/api/admin_api.go:420]: again maybe check for custom error if nothing was updated, otherwise return  err

**Review by nokacper24 — CHANGES_REQUESTED** [review]

File: internal/web_server/web/handlers/api/admin_api.go (hunk 10 above)
Comment (nokacper24) [reply internal/web_server/web/handlers/api/admin_api.go:403]: unless youre replacing content with nothing, then htmx may require 200 ok

File: internal/web_server/web/handlers/api/admin_api.go (hunk 5 above)
Comment (z-rayc) [reply internal/web_server/web/handlers/api/admin_api.go:328]: If there are no rows found, it means the article does not exist in the database already. That is allowed. We will then simply add it to the database. That is why I check for any error beside `sql.errNoRows`. I think it should be fine. Although, the logic here is a bit complex and I do wish to simplify it.

The general flow is:
- Is there an article URL?
  - No: Do nothing, ignore rest of the steps
  - Yes: Check if it already exists
- Does article already exist?
  - No: Find the article data and add it to DB
  - Yes: Continue to next step
- Does the quiz already have the article?
  - No: Add it to the quiz
  - Yes: Continue to next step
- Add article to question and add question to DB.

Comment (z-rayc) [reply internal/web_server/web/handlers/api/admin_api.go:328]: Actually, given that currently user cannot add new articles via "Create/edit question" window (that was not part of original plan, and changed recently), only articles that already exist in quiz, I can skip the "Does article already exist" and "Does the quiz already have the article" steps. Would that be preferable?

File: internal/web_server/web/handlers/api/admin_api.go (hunk 6 above)
Comment (z-rayc) [reply internal/web_server/web/handlers/api/admin_api.go:336]: Unsure. It is their responsibility to check that the article URL is valid. But being unable to fetch data could also be due to server error. The only way to check if the URL is valid is to attempt fetching the data though.

Comment (z-rayc) [reply internal/web_server/web/handlers/api/admin_api.go:336]: I guess I can check what error `GetSmpArticleByURL` returns. Will make my own error types.

**Review by nokacper24 — APPROVED** [review]

**Review by nokacper24 — APPROVED** [review]

**Pull request merged**